include/crab/domains/object_domain.hpp

```diff
@@ -0,0 +1,1303 @@
+                const allocation_site &as) override {
+    crab::CrabStats::count(domain_name() + ".count.ref_make");
+    crab::ScopedCrabStats __st__(domain_name() + ".ref_make");
+
+    ERROR_IF_NOT_REGION(rgn, __LINE__);
+    ERROR_IF_NOT_REF(ref, __LINE__);
+
+    if (is_bottom()) {
+      return;
+    }
+
+    variable_t rep = get_object_rep_by_rgn(rgn);
+    auto old_obj_info = m_obj_info_env[rep];
+    const small_range &num_refs = old_obj_info.count_dom();
+
+    if (num_refs.is_one()) {
+      // if the counting is already one, use odi map in the future
+      move_singleton_to_odi_map(rep);
+    }
+
+    m_obj_info_env.set(rgn,
+                       region_domain_impl::region_info(
+                           old_obj_info.count_dom().increment(),
+                           old_obj_info.init_dom(), old_obj_info.type_dom()));
+
+    CRAB_LOG("object-make", crab::outs()
+                                << "After ref_make(" << ref << "," << rgn << ":"
+                                << rgn.get_type() << "," << size << "," << as
+                                << ")=" << *this << "\n";);
+  }
+
+  // Read the content of reference ref within rgn. The content is
+  // stored in res.
+  void ref_load(const variable_t &ref, const variable_t &rgn,
+                const variable_t &res) override {
+    crab::CrabStats::count(domain_name() + ".count.ref_load");
+    crab::ScopedCrabStats __st__(domain_name() + ".ref_load");
+
+    ERROR_IF_NOT_REGION(rgn, __LINE__);
+    ERROR_IF_ARRAY_REGION(rgn, __LINE__);
+    ERROR_IF_NOT_REF(ref, __LINE__);
+    if ((rgn.get_type().is_bool_region() && !res.get_type().is_bool()) ||
+        (rgn.get_type().is_integer_region() && !res.get_type().is_integer()) ||
+        (rgn.get_type().is_real_region() && !res.get_type().is_real()) ||
+        (rgn.get_type().is_reference_region() &&
+         !res.get_type().is_reference())) {
+      CRAB_ERROR(domain_name(), "::ref_load: type of lhs ", res, " (",
+                 res.get_type(), ") is not compatible with region ", rgn, " (",
+                 rgn.get_type(), ")");
+    }
+
+    if (is_bottom()) {
+      return;
+    }
+
+    if (is_null_ref(ref).is_true()) {
+      CRAB_LOG("object-load", CRAB_WARN(domain_name(), "::ref_load: reference ",
+                                        ref, " is null."););
+      // set_to_bottom();
+      crab::CrabStats::count(domain_name() +
+                             ".count.ref_load.skipped.null_reference");
+      m_base_dom -= res;
+      return;
+    }
+
+    variable_t rep = get_object_rep_by_rgn(rgn);
+    // retrieve an abstract object info
+    auto obj_info = m_obj_info_env[rep];
+
+    const small_range &num_refs = obj_info.count_dom();
+    if (num_refs.is_zero()) {
+      m_base_dom -= res;
+    } else if (num_refs.is_one()) { // singleton object
+      // strong read
+      m_base_dom.assign(res, rgn);
+    } else {
+      // read from odi map
+      // weak read
+      auto obj_dom = m_odi_map[rep];
```

Comment:
this is better.

Here, and everywhere, you should be very careful with copy-assignments. I believe that 

```
auto obj_dom = ...
```
is very expensive and should be
```
auto &obj_dom = ...
```

However, you sometimes re-assign `obj_dom`, so you can't use a reference all the time, but should only copy when necessary.